template/koa-rest-api/src/framework/validation.ts

```diff
@@ -29,13 +31,31 @@ import type { Context } from 'src/types/koa';
  * { "/advertiserId": "advertiserId is required in the URL" }
  * ```
  */
-const parseInvalidFieldsFromError = ({
-  errors,
-}: z.ZodError): Record<string, string> =>
-  Object.fromEntries(
-    errors.map((err) => [`/${err.path.join('/')}`, err.message]),
+
+const parseInvalidFieldsFromError = ({ errors }: z.ZodError) => {
+  const invalidFields: InvalidFields = {};
+  const flattenedIssues: Array<[z.ZodIssue, UnionErrorIdx]> = errors.map(
+    (issue) => [issue, undefined],
   );
 
+  for (const [issue, unionIdx] of flattenedIssues) {
+    if (issue.code === ZodIssueCode.invalid_union) {
+      issue.unionErrors.forEach((err, idx) => {
```

Comment:
Does this work for nested unions?

Reply:
Updated in the [commit](https://github.com/seek-oss/skuba/pull/1339/commits/4402bc80463800e396e5299f54bdb1c5c05aa330)

Reply:
@72636c I have added the logic to handle nested unions. Could you please have a look?🙏

Reply:
🙇

I tried re-implementing the function to understand how it worked, particularly as I was about nervous about `push`ing into an array that was being iterated on. I ended up looking at the following diff; do you think this could be useful to more closely reflect the shape of the Zod schema, i.e. where a union sits "before" the `description` property?

```diff
Snapshot name: `validate blocks invalid nested union prop 1`

-     "/description~union0/content~union1": "Required",
-     "/description~union0~union0": "Expected string, received object",
-     "/id~union0": "Expected string, received null",
-     "/id~union1": "Expected number, received null",
-     "/summary~union1": "Required",
+     "~union0/description~union0": "Expected string, received object",
+     "~union0/description~union1/content": "Required",
+     "~union0/id": "Expected string, received null",
+     "~union1/id": "Expected number, received null",
+     "~union1/summary": "Required",
```

```typescript
const parseInvalidFieldsFromError = (err: z.ZodError): InvalidFields =>
  Object.fromEntries(parseTuples(err, {}));

const parseTuples = (
  { errors }: z.ZodError,
  unions: Record<number, number[]>,
): Array<[string, string]> =>
  errors.flatMap((issue) => {
    if (issue.code === ZodIssueCode.invalid_union) {
      return issue.unionErrors.flatMap((err, idx) =>
        parseTuples(err, {
          ...unions,
          [issue.path.length]: [...(unions[issue.path.length] ?? []), idx],
        }),
      );
    }

    const path = ['', ...issue.path]
      .map((prop, idx) =>
        [
          prop,
          ...(unions[idx] ?? []),
        ].join('~union'),
      )
      .join('/');

    return [[path, issue.message]] as const;
  });
```

Alternatively I can get it to behave identically to yours with a few tweaks:

```diff
const parseTuples = (
  { errors }: z.ZodError,
  unions: Record<number, number[]>,
): Array<[string, string]> =>
  errors.flatMap((issue) => {
    if (issue.code === ZodIssueCode.invalid_union) {
      return issue.unionErrors.flatMap((err, idx) =>
        parseTuples(err, {
          ...unions,
          [issue.path.length]: [...(unions[issue.path.length] ?? []), idx],
        }),
      );
    }

-    const path = ['', ...issue.path]
+    const path = issue.path
      .map((prop, idx) =>
        [
          prop,
          ...(unions[idx] ?? []),
+         ...((idx + 1 === issue.path.length && unions[idx + 1]) || []),
        ].join('~union'),
      )
      .join('/');

-   return [[path, issue.message]] as const;
+   return [[`/${path}`, issue.message]] as const;
  });
```

Reply:
Actually, I think your approach output is more readable.
Would you like to push your changes? or I can update and push.

Reply:
Happy for you to decide which is best here and push 👍. Thanks for your patience!

Reply:
Updated